text/0000-unsized-expr.md

```diff
@@ -0,0 +1,216 @@
+
+```rust
+let x: Box<X> = box f();
+let mut y: Vec<X> = box f();
+
+// Appends the result of `f()` to `y`
+in y.append() { f() }
+```
+
+Note that the last line will be faster than `y.push(f())` because it passes a
+pre-allocated slot in `y` to `f` so that we avoid a copy.
+
+With this protocol in place, we can extend the general idea of "never box return
+values" to unsized types. This RFC introduces the following syntax:
+
+```rust
+fn f() -> [u8] { /* ... */ }
+```
+
+Which can be used like this:
+
+```rust
+let x: Box<[u8]> = box f();
+let mut y: Vec<u8> = box f();
+
+// Appends the result of `f()` to `y`
+in y.append() { f() }
+```
+
+Similarly for traits:
+
+```rust
+fn f() -> fmt::Debug { /* ... */ }
+
+let x: Box<fmt::Debug> = box f();
+```
+
+Note that the return value of `f()` is not restricted to a single implementation
+of `fmt::Debug`, i.e., the following code is valid:
+
+```rust
+fn f(flag: bool) -> fmt::Debug {
+    if flag {
+        *"hello"
+    } else {
+        1234
+    }
+}
+```
+
+Here we also used the implicit conversions `str -> fmt::Debug` and
+`i32 -> fmt::Debug`.
+
+(Note that this concept is orthogonal to the proposed `f() -> impl Trait`
+syntax. Functions declared with an `impl Trait` return type will return one
+concrete implementation of the trait and the return value can be stored on the
+stack.)
+
+This gives the user much greater control over allocations. The user can decide
+freely if they want a `Trait` to be returned in a `Rc<Trait>`, `Arc<Trait>`, or
+`Box<Trait>`. They can also decide if they want a slice to be appended to a
+vector or if they want a completely new vector to be allocated.
+
+## New concepts
+
+It's also interesting to see that this syntax allows us to express new concepts
+in safe Rust that could previously only be expressed with unsafe code.
+
+Consider the following concept:
+
+** The class of types that can be turned into a slice **
+
+With the proposed syntax, this can be expressed as follows:
+
+```rust
+trait IntoSlice<T> {
+    fn into_slice(self) -> [T];
+}
+
+impl<'a, T: Copy> IntoSlice<T> for &'a [T] {
+    fn into_slice(self) -> [T] { self[..] }
+}
+
+impl<T> IntoSlice<T> for Box<[T]> {
+    fn into_slice(self) -> [T] { *self }
+}
+```
+
+It is not possible to have one trait that encapsulates both of these
+implementations safely without the `[T]` syntax. Let's see why:
+
+First of all, since we want the trait to be implemented for all `Box<[T]>`, not
+just those `T` that implement `Copy`, the trait has to take `self` by value.
+Since the return value has to work for `&'a [T]`, the return value has to be
+`&'a [T]`. Therefore the trait has to look like this:
+
+```rust
+trait IntoSlice<'a, T> {
+    fn into_slice(self) -> &'a [T];
+}
+```
+
+But this cannot be implemented for `Box<[T]>` because we either leak the heap
+allocation or return a reference to an already deallocated slice.
+
+(For similar reasons it's not possible to implement the trait on
+`&mut Box<[T]>`)
```

Comment:
For posterity: discussion on IRC lead to clarification that `IntoSlice` is for translating into a slice which *owns* its contents, so it certainly cannot be implemented for `&Box<[T]>`, but it can be implemented with `&mut Box<[T]>` with `mem::replace`.